src/bip340.rs

```diff
@@ -0,0 +1,476 @@
+    #[inline]
+    pub fn as_ptr(&self) -> *const ffi::XOnlyPublicKey {
+        &self.0 as *const _
+    }
+
+    /// Obtains a raw mutable pointer suitable for use with FFI functions
+    #[inline]
+    pub fn as_mut_ptr(&mut self) -> *mut ffi::XOnlyPublicKey {
+        &mut self.0 as *mut _
+    }
+
+    /// Creates a new BIP340 public key from a secret key.
+    #[inline]
+    pub fn from_secret_key<C: Signing>(secp: &Secp256k1<C>, sk: &SecretKey) -> PublicKey {
+        let mut keypair = ffi::KeyPair::new();
+        let mut xonly_pk = ffi::XOnlyPublicKey::new();
+        let mut pk_parity = 0;
+        unsafe {
+            let mut ret = ffi::secp256k1_keypair_create(secp.ctx, &mut keypair, sk.as_c_ptr());
+            debug_assert_eq!(ret, 1);
+            ret =
+                ffi::secp256k1_keypair_xonly_pub(secp.ctx, &mut xonly_pk, &mut pk_parity, &keypair);
+            debug_assert_eq!(ret, 1);
+        }
+        PublicKey(xonly_pk)
+    }
+
+    /// Creates a BIP340 public key directly from a slice
+    #[inline]
+    pub fn from_slice(data: &[u8]) -> Result<PublicKey, Error> {
+        if data.is_empty() || data.len() != constants::BIP340_PUBLIC_KEY_SIZE {
+            return Err(InvalidPublicKey);
+        }
+
+        let mut pk = ffi::XOnlyPublicKey::new();
+        unsafe {
+            if ffi::secp256k1_xonly_pubkey_parse(
+                ffi::secp256k1_context_no_precomp,
+                &mut pk,
+                data.as_c_ptr(),
+            ) == 1
+            {
+                Ok(PublicKey(pk))
+            } else {
+                Err(InvalidPublicKey)
+            }
+        }
+    }
+
+    #[inline]
+    /// Serialize the key as a byte-encoded pair of values. In compressed form
+    /// the y-coordinate is represented by only a single bit, as x determines
+    /// it up to one bit.
+    pub fn serialize(&self) -> [u8; constants::BIP340_PUBLIC_KEY_SIZE] {
+        let mut ret = [0; constants::BIP340_PUBLIC_KEY_SIZE];
+
+        unsafe {
+            let err = ffi::secp256k1_xonly_pubkey_serialize(
+                ffi::secp256k1_context_no_precomp,
+                ret.as_mut_c_ptr(),
+                self.as_c_ptr(),
+            );
+            debug_assert_eq!(err, 1);
+        }
+        ret
+    }
+}
+
+impl CPtr for PublicKey {
+    type Target = ffi::XOnlyPublicKey;
+    fn as_c_ptr(&self) -> *const Self::Target {
+        self.as_ptr()
+    }
+
+    fn as_mut_c_ptr(&mut self) -> *mut Self::Target {
+        self.as_mut_ptr()
+    }
+}
+
+/// Creates a new BIP340 public key from a FFI x-only public key
+impl From<ffi::XOnlyPublicKey> for PublicKey {
+    #[inline]
+    fn from(pk: ffi::XOnlyPublicKey) -> PublicKey {
+        PublicKey(pk)
+    }
+}
+
+serde_impl_from_slice!(PublicKey);
+
+impl<C: Signing> Secp256k1<C> {
+    /// Create a BIP340 signature.
+    pub fn bip340_sign(
+        &self,
+        msg: &Message,
+        sk: &SecretKey,
+        aux_rand: &[u8; 32],
```

Comment:
IMO this should be splitted to 2 functions:
```rust
bip340_sign(&self, &Message, &SecretKey)
bip340_sign_randomize(&self, &Message, &SecretKey, rng: impl Rng + CryptoRng)
```
or something like that

Reply:
I agree it's nice to have some convenience functions. I defined three now:
```rust
bip340_sign(&self, msg: &Message, sk: &SecretKey)
```
Internally uses `OsRng` to generate the auxiliary random data. Requires `rand-std` feature (I hope I understood the purpose of this feature correctly)

```rust
bip340_sign_with_rng<R: Rng + CryptoRng>(
        &self,
        msg: &Message,
        sk: &SecretKey,
        rng: &mut R,
    )
```
Uses the given rng to generate the auxiliary random data. Requires `rand` feature.

```rust
bip340_sign_with_aux_rand(
        &self,
        msg: &Message,
        sk: &SecretKey,
        aux_rand: &[u8; 32],
    )
```
Uses the provided auxiliary random data, doesn't require any special feature. I think it's nice to keep that as it makes it possible to create BIP340 signatures without compiling with "rand" (also practical for testing, but could make it crate pub if you think it's better not to have it public).

I hope that matches with what you had in mind.

Reply:
I would just like to point out that this is diverging from out usual API (not saying that's necessarily bad), our usuall API takes an `Rng` for random, and in the test we use fake Rng's. 
Personally I don't mind too much. @apoelstra what do you think?

Reply:
Ah maybe I have misunderstood how you wanted the function `bip340_sign`. As you didn't put any rng parameter there nor auxiliary random data I assumed you wanted it to be instantiated inside the function, but how were you thinking about it? I can just drop the `bip340_sign` function if instantiating an rng within the library is not desired.

> in the test we use fake Rng's.

I saw quite some tests using `thread_rng` for key pair generation so I thought I would use that as well, but if you think it's better to have something deterministic I can replace maybe with `DumbRng`?

Reply:
> I saw quite some tests using thread_rng for key pair generation so I thought I would use that as well, but if you think it's better to have something deterministic I can replace maybe with DumbRng?

That depends on what you want to test, you asked for a way to test with a specific aux_rand

> Ah maybe I have misunderstood how you wanted the function bip340_sign. As you didn't put any rng parameter there nor auxiliary random data I assumed you wanted it to be instantiated inside the function, but how were you thinking about it?

I was thinking about not using aux rand there, and passing NULL

Reply:
> That depends on what you want to test, you asked for a way to test with a specific aux_rand

Got you, indeed I can get rid of the `_with_aux_rand` version if you don't think it's useful for users and replace the tests with a dumb rng.